dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+```haskell
+unify a@Match {} b = die $ ProjectionL a b
+unify a b@Match {} = die $ ProjectionR a b
+
+unify a@Get {} b = die $ ProjectionL a b
+unify a a@Get {} = die $ ProjectionR a b
+```
+
+The `let`-expression of both kinds should not enter unifier at all. They can always be normalised to non-`let`-expressions.
+```haskell
+unify Let {} _ = error "unifier: left argument is a let-expression"
+unify _ Let {} = error "unifier: right argument is a let-expression"
+
+unify LetRec {} _ = error "unifier: left argument is a let-expression"
+unify _ LetRec {} = error "unifier: right argument is a let-expression"
+```
+
+And the last case is fired after all other fail, which means that types don't match.
+```haskell
+unify a b = die $ Mismatch a b
+```
+
+## Normalization
+
+Before we can start with inference, there is another matter to discuss. In simpler type systems the type are always in their normal form, there is no abstraction that disobeys `f x ~ g x -> f ~ g` rule, which means that there is no type lambdas.
+
+We have no such restrictions here. There _can_ be lambdas, which gives us a need to evaluate type expression before using them in type checks.
+
+However, the normalization is easy, since the expression to be normalised will be typechecked _before_ it enters normaliser, which kinda gives us moral right to ignore all possible "runtime" failures in normalisation.
+
+I will mostly skip the description of it. The single problem I met is a representation of types. I made a separate constructor for that:
+```haskell
+Axiom :: Name -> self -> AST_ self
+```
+This constructor can only be inserted by normalizer, by evaluating, for instance
+```haskell
+let
+  data Bool where
+    | True  : Bool
+    | False : Bool
+in
+  <ctx>
+```
+
+The names `Bool`, `True` and `False` will be replaced in `ctx` with (in pseudosyntax)
+
+```
+axiom "True" Bool
+axiom "True" Bool
+axiom "Bool" Type
+```
+
+accordingly.
+
+## Type Inference
+
+Again, the inference will consist of 2 functions: first one does the work, second one calls the first one _and then applies all substitutions to the program_. Why would we do that? Well, after we finish, there still can be free variables in the program, which means that perhaps some types are not fixed, and this is might cause trouble in the codegen phase.
+
+The only thing is missing is a typing context, which is usually denoted as capital greek Gamma letter. In our case, it will be a map between names and their types.
+
+We will now look at the first function of the two:
+
+```haskell
+infer
+  :: ( Unifies m
+     , HasSubst m
+     , HasContext m
+     , Refreshes m
+     )
+  => AST
+  -> Sem m AST
+```
```

Comment:
What is `Sem`?

Reply:
An implementation detail - https://hackage.haskell.org/package/polysemy-1.6.0.0/docs/Polysemy.html#t:Sem